rust/sealevel/programs/interchain-gas-paymaster/src/accounts.rs

```diff
@@ -0,0 +1,298 @@
+//! Interchain gas paymaster accounts.
+
+use std::{cmp::Ordering, collections::HashMap};
+
+use access_control::AccessControl;
+use account_utils::{AccountData, DiscriminatorData, DiscriminatorPrefixed, SizedData};
+use borsh::{BorshDeserialize, BorshSerialize};
+use solana_program::{clock::Slot, program_error::ProgramError, pubkey::Pubkey};
+
+use hyperlane_core::{H256, U256};
+
+use crate::error::Error;
+
+/// The scale for token exchange rates, i.e. a token exchange rate of 1.0 is
+/// represented as 10^19.
+pub const TOKEN_EXCHANGE_RATE_SCALE: u128 = 10u128.pow(19);
+/// The number of decimals for the native SOL token.
+pub const SOL_DECIMALS: u8 = 9;
+
+/// A gas oracle that provides gas data for a remote chain.
+#[derive(BorshSerialize, BorshDeserialize, Debug, PartialEq, Clone)]
+pub enum GasOracle {
+    /// Remote gas data stored directly in the variant data.
+    RemoteGasData(RemoteGasData),
+    // Future gas oracle variants could include a Pyth type, generalized CPI type, etc.
+}
+
+impl Default for GasOracle {
+    fn default() -> Self {
+        GasOracle::RemoteGasData(RemoteGasData::default())
+    }
+}
+
+/// The account for the program's global data.
+pub type ProgramDataAccount = AccountData<ProgramData>;
+
+/// A singleton account that stores the program's global data.
+#[derive(BorshSerialize, BorshDeserialize, Debug, PartialEq, Default)]
+pub struct ProgramData {
+    /// The bump seed for the program data PDA.
+    pub bump_seed: u8,
+    /// The number of gas payments made by in the program.
+    pub payment_count: u64,
+}
+
+impl SizedData for ProgramData {
+    fn size(&self) -> usize {
+        // 1 for bump_seed
+        // 8 for payment_count
+        1 + 8
```

Comment:
is sealevel no_std? If it supports std we could use e.g. `std::mem::size_of::<u64>()`. If the compilation target is always the same for sealevel, then we could just directly do `std::mem::size_of::<ProgramData>()`

Reply:
mmm good point

Because these sizes relate to Borsh encoding https://borsh.io/ and not necessarily the memory layout, I think we can only do this for primitive types. Because I haven't used `std::mem::size_of` elsewhere I'm gonna merge as-is for consistency, and then if we have time later to clean things up i'd support moving over primitive type sizes to `std::mem::size_of`